File: src/AbhimoTracker.Employee/Services/UploadQueue.cs
using System.IO;
using System.Text.Json;
using System.Timers;
using AbhimoTracker.Employee.Models;
using Timer = System.Timers.Timer;

namespace AbhimoTracker.Employee.Services;

/// <summary>
/// Disk-backed queue for activity events. Port of desktop-tray/upload-queue.js.
/// Batches entries instead of sending one request per event, and survives a
/// network blip, backend restart, or agent restart without losing tracked
/// time -- queued items persist to a local JSON file and are retried on the
/// next flush ("stores data locally temporarily if internet disconnects").
///
/// Each item carries a stable client_batch_id (GUID) generated at enqueue
/// time, so a retry after a lost response can be deduped server-side
/// (ActivityController::ingest's `ON DUPLICATE KEY UPDATE id = id`) instead
/// of double-counting duration.
/// </summary>
public sealed class UploadQueue : IDisposable
{
    private readonly string _filePath;
    private readonly int _maxBatch;
    private readonly TimeSpan? _maxAge;
    private readonly Timer _timer;
    private readonly object _lock = new();
    private List<ActivityEntry> _items;
    private Func<IReadOnlyList<ActivityEntry>, CancellationToken, Task<bool>>? _sender;
    private bool _flushing;

    public int PendingCount { get { lock (_lock) return _items.Count; } }

    public UploadQueue(string userDataDir, int flushIntervalMs, int maxBatch, TimeSpan? maxAge)
    {
        Directory.CreateDirectory(userDataDir);
        _filePath = Path.Combine(userDataDir, "activity-queue.json");
        _maxBatch = maxBatch;
        _maxAge = maxAge;
        _items = Load();
        _timer = new Timer(flushIntervalMs) { AutoReset = true };
        _timer.Elapsed += async (_, _) => await FlushAsync();
    }

    public void SetSender(Func<IReadOnlyList<ActivityEntry>, CancellationToken, Task<bool>> sender) => _sender = sender;

    public void Enqueue(ActivityEntry item)
    {
        item.QueuedAtUnixMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        lock (_lock) _items.Add(item);
        Persist();
    }

    public void Start() => _timer.Start();
    public void Stop() => _timer.Stop();

    public async Task FlushAsync(CancellationToken ct = default)
    {
        if (_flushing || _sender is null) return;
        List<ActivityEntry> batch;
        lock (_lock)
        {
            if (_flushing) return;
            _flushing = true;
        }
        try
        {
            DropExpired();
            lock (_lock)
            {
                if (_items.Count == 0) return;
                batch = _items.Take(_maxBatch).ToList();
            }

            bool ok;
            try
            {
                ok = await _sender(batch, ct);
            }
            catch
            {
                ok = false;
            }

            if (ok)
            {
                lock (_lock) _items = _items.Skip(batch.Count).ToList();
                Persist();
            }
            // On failure the batch stays at the front of the queue and is
            // retried next cycle -- no data loss, just delay.
        }
        finally
        {
            lock (_lock) _flushing = false;
        }
    }

    /// <summary>
    /// Without this, a batch the server will reject forever (timestamps fell
    /// outside the accepted backdate window while offline) sits at the front
    /// of the queue and blocks every entry queued behind it indefinitely --
    /// same reasoning as upload-queue.js's _dropExpired().
    /// </summary>
    private void DropExpired()
    {
        if (_maxAge is not { } maxAge) return;
        var cutoff = DateTimeOffset.UtcNow - maxAge;
        lock (_lock)
        {
            var before = _items.Count;
            _items = _items.Where(i => i.Timestamp.ToUniversalTime() >= cutoff.UtcDateTime).ToList();
            if (_items.Count != before) Persist();
        }
    }

    private List<ActivityEntry> Load()
    {
        try
        {
            if (File.Exists(_filePath))
            {
                var json = File.ReadAllText(_filePath);
                return JsonSerializer.Deserialize<List<ActivityEntry>>(json) ?? [];
            }
        }
        catch
        {
            // Corrupt queue file -- start fresh rather than crashing.
        }
        return [];
    }

    private void Persist()
    {
        try
        {
            List<ActivityEntry> snapshot;
            lock (_lock) snapshot = [.. _items];
            File.WriteAllText(_filePath, JsonSerializer.Serialize(snapshot));
        }
        catch
        {
            // Best-effort persistence -- next successful flush will retry writing anyway.
        }
    }

    public void Dispose()
    {
        _timer.Stop();
        _timer.Dispose();
    }
}

File: desktop-tray/upload-queue.js
'use strict';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

/**
 * Disk-backed queue for activity events (app segments, idle/active
 * transitions). Batches them instead of sending one request per event, and
 * survives a network blip, backend restart, or agent restart without
 * losing tracked time -- queued items persist to disk and are retried on
 * the next flush.
 *
 * Each item gets a stable client_batch_id (uuid) generated at enqueue
 * time, so if a request actually succeeded on the server but the response
 * was lost (e.g. connection dropped after the write), a retry can be
 * deduped server-side instead of double-counting duration. The backend
 * needs to honor this id as an idempotency key.
 */
class UploadQueue {
  constructor(userDataDir, { flushIntervalMs = 30000, maxBatch = 25, maxAgeMs = null } = {}) {
    this.filePath = path.join(userDataDir, 'activity-queue.json');
    this.flushIntervalMs = flushIntervalMs;
    this.maxBatch = maxBatch;
    // Items older than this are dropped instead of ever being sent. Left
    // null by default (never expire) so any other future use of this
    // generic queue isn't forced to opt in. See _dropExpired() below for
    // why this exists.
    this.maxAgeMs = maxAgeMs;
    this._items = this._load();
    this._timer = null;
    this._sender = null; // async (items) => boolean
    this._flushing = false;
  }

  /** @param {(items: object[]) => Promise<boolean>} sender */
  setSender(sender) {
    this._sender = sender;
  }

  enqueue(item) {
    this._items.push({ ...item, client_batch_id: crypto.randomUUID(), queued_at: Date.now() });
    this._persist();
  }

  start() {
    if (this._timer) return;
    this._timer = setInterval(() => this.flush(), this.flushIntervalMs);
  }

  stop() {
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
    }
  }

  async flush() {
    if (this._flushing || !this._sender || this._items.length === 0) return;
    this._flushing = true;
    try {
      this._dropExpired();
      if (this._items.length === 0) return;

      const batch = this._items.slice(0, this.maxBatch);
      let ok = false;
      try {
        ok = await this._sender(batch);
      } catch (_) {
        ok = false;
      }
      if (ok) {
        this._items = this._items.slice(batch.length);
        this._persist();
      }
      // On failure the batch stays at the front of the queue and is
      // retried next cycle -- no data loss, just delay.
    } finally {
      this._flushing = false;
    }
  }

  /**
   * Without this, a batch the server will reject forever (its timestamps
   * fell outside the backend's accepted backdate window while the agent
   * was offline) sits at the front of the queue and is retried every
   * cycle, indefinitely blocking every entry queued behind it -- a laptop
   * left off for a long weekend could silently stop reporting *any* new
   * activity even after reconnecting, since flush() always tries the
   * batch at index 0 first.
   *
   * Runs on every flush() (not just once) since staleness is relative to
   * "now" -- an item queued when fresh can still age past maxAgeMs while
   * merely waiting its turn in a backlog, not only while offline.
   * Unparsable timestamps are kept rather than guessed away -- the server
   * is the actual authority on whether an entry is acceptable, so let a
   * real rejection response explain why instead of silently dropping
   * something this method can't confidently judge.
   */
  _dropExpired() {
    if (!this.maxAgeMs) return;
    const cutoff = Date.now() - this.maxAgeMs;
    const before = this._items.length;
    this._items = this._items.filter((item) => {
      const t = new Date(item.timestamp).getTime();
      return Number.isNaN(t) || t >= cutoff;
    });
    if (this._items.length !== before) {
      this._persist();
      console.warn(`UploadQueue: dropped ${before - this._items.length} entries older than the server's accepted window.`);
    }
  }

  _load() {
    try {
      if (fs.existsSync(this.filePath)) {
        const parsed = JSON.parse(fs.readFileSync(this.filePath, 'utf8'));
        return Array.isArray(parsed) ? parsed : [];
      }
    } catch (err) {
      console.error('Activity queue file was unreadable, starting fresh:', err.message);
    }
    return [];
  }

  _persist() {
    try {
      fs.writeFileSync(this.filePath, JSON.stringify(this._items), { mode: 0o600 });
    } catch (err) {
      console.error('Failed to persist activity queue:', err.message);
    }
  }

  get pendingCount() {
    return this._items.length;
  }
}

module.exports = { UploadQueue };
